atef/widgets/archive_viewer.py

```diff
@@ -0,0 +1,619 @@
+            """
+            def fn():
+                now = datetime.datetime.today()
+                prev = now - datetime.timedelta(seconds=s)
+                self.time_plot.setXRange(prev.timestamp(), now.timestamp())
+                self.time_plot.updateXAxis()
+            return fn
+
+        self.button_day.clicked.connect(_set_time_span_fn(24*60*60))
+        self.button_week.clicked.connect(_set_time_span_fn(24*60*60*7))
+        self.button_month.clicked.connect(_set_time_span_fn(24*60*60*30))
+
+    def _setup_pv_selector(self) -> None:
+        """
+        Set up pv validation on input field line edit.
+        Attempts to pass pv name to ``ArchiveViewerWidget.add_signal()``
+        """
+        # validate PV form, returnPressed unless valid
+        regexp = QRegularExpression(r'^\w+(:\w+)+(\.\w+)*$')
+        validator = QRegularExpressionValidator(regexp)
+        self.input_field.setValidator(validator)
+
+        def _add_item():
+            """ slot for input_field submission """
+            # grab and clear text
+            pv = self.input_field.text()
+
+            # add item
+            self.add_signal(pv)
+
+        self.input_field.returnPressed.connect(_add_item)
+
+    def _update_curves(self):
+        """ Clears the timeplot and adds any PV's present in the PVModel """
+        # grab all the list items
+        pv_data = self.curve_list.model().pvs
+
+        # Re-add curves
+        self.time_plot.clearCurves()
+        for pv in pv_data:
+
+            data = self.get_pv_data_snippet(pv[0])
+
+            self.time_plot.addYChannel(
+                y_channel=f'ca://{pv[0]}',
+                name=f'{pv[0]} ({data["meta"].get("EGU","")})',
+                symbol=pv[1]['symbol'],
+                color=pv[1]['color'],
+                lineStyle=pv[1]['lineStyle'],
+                lineWidth=pv[1]['lineWidth'],
+                useArchiveData=True,
+                yAxisName='yAxis'
+            )
+
+        try:
+            self.time_plot.setLabel('yAxis', text='')
+        except Exception:
+            # pyqtgraph raises a vanilla exception
+            # if a better way to find the left axis name exists, use it
+            logger.debug('left axis does not exist to rename')
+
+        self.time_plot.setShowLegend(True)
+
+    def add_signal(self, pv: str, dev_attr: Optional[str] = None) -> None:
+        """
+        Adds a PV to the ArchiverViewerWidget's PVModel
+        Ensures PV's have at least 3 data points in the archiver before
+        adding a PV to the model
+
+        Updates the time plot widget and clears the input field if
+        successful
+
+        Parameters
+        ----------
+        pv : str
+            the PV to be added (eg. MR2L0:RTD:1:TEMP)
+        dev_attr : Optional[str], optional
+            the ophyd attribute name corresponding to the ``pv``,
+            by default None
+        """
+        # check if data exists
+        data = self.get_pv_data_snippet(pv)
```

Comment:
I clicked on the awesome new "view archive data" button in the PV configuration comparison of `lfe.json`, and it's doing what I've asked. However, the archive requests take a long time and are run in the GUI thread, meaning it blocks everything else happening in atef until these things are run.  Perhaps it should be sent to a thread, instead? 

In the case of a slow archiver response over ssh, this is actually taking well over a minute (maybe two at this point? I started typing the comment when I noticed it, and it still hasn't finished)... 

Edit: hmm, it's still going. I have a slight modification of archapp that [indicates this](https://gist.github.com/klauer/533197dbf3698353beb7bfed682a17ad) for a single click. That seems like way too many requests. What's going on here?

Reply:
This is weird... I only get one call to the archapp `get_raw` method whenever I draw a curve, as I expect.  That was for a single click of which button?  The "view archive data" button on the PV comparison page?

Your output definitely shows too many requests (multiple calls to a single PV), but I can't reproduce it from psbuild.  I wonder if archapp is retrying failed requests... it doesn't seem to be from my reading of the code...

All that being said it's probably worth tossing in a thread.  👍

Reply:
archapp v1.1.0
`python -m atef config atef/tests/configs/lfe.json`
On "Good vacuum", click the archive viewer button:
<img width="449" alt="image" src="https://user-images.githubusercontent.com/5139267/202546144-7929a127-5e95-4bbd-8ccf-5a0e1d4cfb69.png">
This may be retries-related. Some of this may still be a me-problem, as `curl` -ing those URLs locally works without issue, but archapp/PyDM seem to take issue with them

Reply:
Ken astutely noticed that this was querying each PV N^2 times, where N was the number of PV's being added at once.  This was fixed by not updating the plot with every PV added, rather updating it all at the end

The calls to archapp were also cached.  [66a1923](https://github.com/pcdshub/atef/pull/132/commits/66a1923dbb540372fed0224df42369541443813b)

I'm looking into shoving the calls into worker threads now as well